RMS_MGMT

yes1n opened the pull request:
https://github.com/glpi-project/glpi-agent/discussions/495 I did a pull request for now because it works but needs improvement. Added rms.t and rms.reg for the test.

yes1n committed:
RMS_MGMT

https://github.com/glpi-project/glpi-agent/discussions/495

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,90 @@
+                maxdepth    => 1,
+                logger => $params{logger}
+            );
+        }
```
As I said in #495, don't duplicate code that does exactly the same thing and will return finally nothing. With your explanation, I can assume the first call is sufficient in all case.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,90 @@
+            );
+        }
+
+        return $key && (keys %$key);
```
Just return true here, the purpose of this API is just to know if the agent should run the RMS inventory.
```suggestion
        return 1;
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,90 @@
+    my $InternetID = _getID(
+        osname  => OSNAME,
+        logger  => $logger
+    );
```
As I said in #495, don't pass osname but directly make a test to check which API to run.
```suggestion
    my $InternetID = OSNAME eq 'MSWin32' ? _getID_MSWin32(logger  => $logger) : _getID(logger  => $logger);
    );
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,90 @@
+    my (%params) = @_;
+    my $osname = delete $params{osname} // '';
+    return _getID_MSWin32() if $osname eq "MSWin32";
+}
```
As I said in #495, we propably need here to analyse `rms` command output in not `MSWin32` case. And you'll probably have to provide some output to run test on and the related test into the `rms.t` file.

g-bougard commented on t/tasks/inventory/generic/remote_mgmt/rms.t:
```diff
@@ -0,0 +1,48 @@
+        '_getRegistryKey',
+        _mockGetRegistryKey($test)
+    );
+    my $internetID = GLPI::Agent::Task::Inventory::Generic::Remote_Mgmt::RMS::_getID(osname => "MSWin32");
```
To follow my suggestions, just fix the call here to support `MSWin32` case test.
```suggestion
    my $internetID = GLPI::Agent::Task::Inventory::Generic::Remote_Mgmt::RMS::_getID_MSWin32();
```

g-bougard commented on t/tasks/inventory/generic/remote_mgmt/rms.t:
```diff
@@ -0,0 +1,48 @@
+    'GLPI::Agent::Tools::Win32'
+);
+
+foreach my $test () {
```
Something is missing here as no test is run.
```suggestion
my %win32_tests = (
    yes1n => "<expected id>",
);

plan tests => (scalar keys %win32_tests) + 1;

foreach my $test (keys(%win32_tests)) {
```
P.S.: I used `yes1n` as key in the list, but as I explained in my last comment below, you may want to better use your rms version as key.

g-bougard commented on t/tasks/inventory/generic/remote_mgmt/rms.t:
```diff
@@ -0,0 +1,48 @@
+
+        # We can mock getRegistryKey or better _getRegistryKey to cover getRegistryValue
+        my $path = $params{path} || $params{keyName};
+        my $file = "resources/generic/rms/rms.reg";
```
Don't limit your test to only hardcoded one test. You should have kept the original code and rename your registry file accordingly.
```suggestion
        my $file = "resources/generic/rms/rms-$test-$path.reg";
```
Following this suggestion and the `%win32_tests` list I defined, you should rename your registry file to `resources/generic/rms/rms-yes1n-InternetId.reg`. The purpose is to be able to add new tests to cover other test cases. I used `yes1n` as key, but you may want to replace it with the current tested rms version.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,90 @@
+
+sub _getID_MSWin32 {
+
+    GLPI::Agent::Tools::Win32->use();
+
+    my $clientid = getRegistryValue(
+        path => 'HKEY_LOCAL_MACHINE/SOFTWARE/Usoris/Remote Utilities Host/Host/Parameters/InternetId',
+    );
+
+    $clientid = hex2dec($clientid);
+
+    $clientid = substr($clientid, 0, index($clientid, '</internet_id'));
+    $clientid = substr($clientid, index($clientid, '<internet_id>')+13);
+
+    return $clientid;
+
+}
```
Here, I suspect you're trying to extract a XML value. You should use `GLPI::Agent::XML` class to extract the data from the expected place in the XML as nothing guaranty you there's only one `internet_id` node.
```suggestion
    GLPI::Agent::Tools::Win32->use();
    GLPI::Agent::XML->use();

    my $internetid = getRegistryValue(
        path => 'HKEY_LOCAL_MACHINE/SOFTWARE/Usoris/Remote Utilities Host/Host/Parameters/InternetId',
    );

    $internetid = hex2dec($clientid);

    my $tree = GLPI::Agent::XML->new(string => $internetid)->dump_as_hash();

    return unless defined($tree) && defined($tree->{nodes_path_to_internet_id});

    return $tree->{nodes_path_to_internet_id}->{internet_id};
}
```
The `return unless ...` here must test all the path nodes, not just the last one. I'll try to tune my suggestion after analysing your registry file.

g-bougard reviewed (changes requested):
Hi @yes1n 
thank you for your PR, anyway I expected to still updated it following the few suggestions I made in the #495 discussion.

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update t/tasks/inventory/generic/remote_mgmt/rms.t

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update t/tasks/inventory/generic/remote_mgmt/rms.t

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update t/tasks/inventory/generic/remote_mgmt/rms.t

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update RMS.pm

#499
https://github.com/glpi-project/glpi-agent/pull/499#discussion_r1336744858

yes1n committed:
Update RMS.pm

#499
https://github.com/glpi-project/glpi-agent/pull/499#discussion_r1336754897

yes1n committed:
#499 rename reg file

#499
https://github.com/glpi-project/glpi-agent/pull/499#discussion_r1336783033

yes1n committed:
Update rms.t

#499
https://github.com/glpi-project/glpi-agent/pull/499#discussion_r1336764552

g-bougard commented on t/tasks/inventory/generic/remote_mgmt/rms.t:
```diff
@@ -0,0 +1,54 @@
+        _mockGetRegistryKey($test)
+    );
+    my $internetID = GLPI::Agent::Task::Inventory::Generic::Remote_Mgmt::RMS::_getID_MSWin32();
+    is($internetID, "RMS win32 getID - $test");
```
You missed to test the expected value:
```suggestion
    is($win32_tests{$test}, $internetID, "RMS win32 getID - $test");
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,89 @@
+    }
+}
+
+# sub _getID {
+#     my (%params) = @_;
+#     my $osname = delete $params{osname} // '';
+#     return _getID_MSWin32() if $osname eq "MSWin32";
+# }
```
Don't comment a method you're using on l.48, this is a syntax error. Just comment the content if you still don't have the code to parse `rms` command output providing the expected id.
```suggestion
sub _getID {
#     my (%params) = @_;
}
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,89 @@
+            maxdepth    => 1,
+            logger => $params{logger}
+        );
+        # if (!$key && is64bit()) {
+        #     $key = getRegistryKey(
+        #         path => 'HKEY_LOCAL_MACHINE/SOFTWARE/Usoris/Remote Utilities Host/Host/Parameters',
+        #         # Important for remote inventory optimization
+        #         required    => [ qw/InternetId/ ],
+        #         maxdepth    => 1,
+        #         logger => $params{logger}
+        #     );
+        # }
```
Just remove this commented code.
```suggestion
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,89 @@
+
+    my $tree = GLPI::Agent::XML->new(string => $internetid)->dump_as_hash();
+
+    return unless defined($tree) && defined($tree->{nodes_path_to_internet_id});
+
+    return $tree->{nodes_path_to_internet_id}->{internet_id};
```
You still need here to find the `internet_id` node the the `$tree` structure.

yes1n replied:
Help me do this, I'm not good at Perl. Here is the xml text from the registry:
```
﻿<?xml version="1.0" encoding="UTF-8"?>
<rms_internet_id_settings version="70170"><internet_id>891-759-358-441</internet_id><use_inet_connection>true</use_inet_connection><inet_server></inet_server><use_custom_inet_server>false</use_custom_inet_server><inet_id_port>5655</inet_id_port><use_inet_id_ipv6>false</use_inet_id_ipv6><inet_id_use_pin>false</inet_id_use_pin><inet_id_pin></inet_id_pin></rms_internet_id_settings>
```

g-bougard replied:
Okay, here is the code I think you should try:
```suggestion
    return unless defined($tree) && defined($tree->{rms_internet_id_settings});

    return $tree->{rms_internet_id_settings}->{internet_id};
```

yes1n committed:
Update t/tasks/inventory/generic/remote_mgmt/rms.t

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n commented:
I'm checking RMS.pm. Doesn't work yet, don't merge yet

yes1n committed:
Update RMS.pm

#499 It works in this combination. Please check the code for correctness. The registry path is the same on both Win32 and Win64

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,73 @@
+        return 1;
+    }
+
+    return canRun('rms');
```
As you reduced support to `MSwin32`, don't check `rms` command is runnable.
```suggestion
    return 0;
```

g-bougard commented on t/tasks/inventory/generic/remote_mgmt/rms.t:
```diff
@@ -0,0 +1,54 @@
+        '_getRegistryKey',
+        _mockGetRegistryKey($test)
+    );
+    my $internetID = GLPI::Agent::Task::Inventory::Generic::Remote_Mgmt::RMS::_getID_MSWin32();
```
The function has been renamed in RMS.pm.
```suggestion
    my $internetID = GLPI::Agent::Task::Inventory::Generic::Remote_Mgmt::RMS::_getID();
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,73 @@
+
+use parent 'GLPI::Agent::Task::Inventory::Module';
+
+use English qw(-no_match_vars);
```
A dependency is missing here.
```suggestion
use English qw(-no_match_vars);
use UNIVERSAL::require;
```
Unless you prefer include the dependencies from l.57 & l.58 in the case you only want to support `MSWin32` (see also my comment below).

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,73 @@
+
+    my $internetid = getRegistryValue(
+        path => 'HKEY_LOCAL_MACHINE/SOFTWARE/Usoris/Remote Utilities Host/Host/Parameters/InternetId',
+    );
```
Just noticed a missing here: `logger` needs to be passed to `getRegistryValue()` call via `%params`
```suggestion
sub _getID {
    my (%params) = @_;

    GLPI::Agent::Tools::Win32->use();
    GLPI::Agent::XML->use();

    my $internetid = getRegistryValue(
        path => 'HKEY_LOCAL_MACHINE/SOFTWARE/Usoris/Remote Utilities Host/Host/Parameters/InternetId',
        %params
    );
```

Also, I'm okay to leave the `...->use()` calls here if you think other platforms support will be provided. But if you think to keep only `MSWin32` support, they can be removed and imported in the header.

g-bougard replied:
@yes1n did you miss this one ? I must admit I failed to add it to the previous review before submitting it.

g-bougard replied:
Also, test is failing as it tries to load `resources/generic/rms/rms-7.0.0.1-Parameters.reg`: so you need to rename `resources/generic/rms/rms-7.0.0.1-RMS.reg` replacing `RMS` by `Parameters` at the end

yes1n committed:
Update t/tasks/inventory/generic/remote_mgmt/rms.t

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

g-bougard commented:
Hi @yes1n 
there's still not resolved comments. Can you check them ?
 1. to fix `logger` not used when calling `getRegistryValue()`
 2. your registry file needs to be renamed

yes1n commented:
Hi, @g-bougard 
1. Regarding the logger, please suggest a change, I will make it. I don't know how to do it
2. Registry file renamed in [7d6759a](https://github.com/glpi-project/glpi-agent/pull/499/commits/7d6759ac4c91c2b3300f01338380d91ef1b767ca)

g-bougard commented:
> Hi, @g-bougard
> 
>     1. Regarding the logger, please suggest a change, I will make it. I don't know how to do it
> 
>     2. Registry file renamed in [7d6759a](https://github.com/glpi-project/glpi-agent/pull/499/commits/7d6759ac4c91c2b3300f01338380d91ef1b767ca)

Please read my pending comments from yesterday where you have the suggestion (`%params` support in `_getID()`) and an explanation on why the registry filename is still not the right.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,74 @@
+sub isEnabled {
+    my (%params) = @_;
+
+    if (OSNAME eq 'MSWin32') {
+
+        GLPI::Agent::Tools::Win32->use();
+
+        my $key = getRegistryKey(
+            path => 'HKEY_LOCAL_MACHINE/SOFTWARE/Usoris/Remote Utilities Host/Host/Parameters',
+            # Important for remote inventory optimization
+            required    => [ qw/InternetId/ ],
+            maxdepth    => 1,
+            logger => $params{logger}
+        );
+
+        return 1;
+    }
+
+    return 0;
```
We missed `$key` is not used here. I would like also simplify the code until you provide a process to check this on other platform:
```suggestion
    return 0 unless OSNAME eq 'MSWin32';

    GLPI::Agent::Tools::Win32->use();

    my $key = getRegistryKey(
        path => 'HKEY_LOCAL_MACHINE/SOFTWARE/Usoris/Remote Utilities Host/Host/Parameters',
        # Important for remote inventory optimization
        required    => [ qw/InternetId/ ],
        maxdepth    => 1,
        logger => $params{logger}
    );

    return defined($key) ? 1 : 0;
```

g-bougard reviewed (changes requested):
Hi @yes1n

my apologies, it seems I missed to validate the comment I was talking about.

Anyway, I found also another little issue in `isEnabled()` function.

Can you can know my requests ?

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n committed:
Rename reg file

#499  https://github.com/glpi-project/glpi-agent/pull/499#discussion_r1337604523

yes1n commented:
@g-bougard, many thanks for your help. Applied your changes and renamed the file to rms-7.0.0.1-Parameters.reg

g-bougard commented on t/tasks/inventory/generic/remote_mgmt/rms.t:
```diff
@@ -0,0 +1,54 @@
+        _mockGetRegistryKey($test)
+    );
+    my $internetID = GLPI::Agent::Task::Inventory::Generic::Remote_Mgmt::RMS::_getID();
+    is($win32_tests{$test}, $internetID, "RMS win32 getID - $test");
```
Sorry, I made a mistake here. This is not critical but the test failure report is wrong as the 2 first parameters on `is()` must be exchanged:
```suggestion
    is($internetID, $win32_tests{$test}, "RMS win32 getID - $test");
```

g-bougard reviewed (changes requested):
Hi @yes1n 

there's a last critical problem with the test as it fails.
Investigating, I found you just have to remove the 3 following lines from your registry file as they are not expected by the mocking API:
```
[HKEY_LOCAL_MACHINE\SOFTWARE\Usoris]

[HKEY_LOCAL_MACHINE\SOFTWARE\Usoris\Remote Utilities Host]

[HKEY_LOCAL_MACHINE\SOFTWARE\Usoris\Remote Utilities Host\Host]

```
So the beginning of the file must look like:
```
Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\SOFTWARE\Usoris\Remote Utilities Host\Host\Parameters]
```

There also a not critical mistake (my fault) in the test file, check my comment.

yes1n committed:
Update t/tasks/inventory/generic/remote_mgmt/rms.t

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n commented:
> Hi @yes1n
> 
> there's a last critical problem with the test as it fails. Investigating, I found you just have to remove the 3 following lines from your registry file as they are not expected by the mocking API:
> 
> ```
> [HKEY_LOCAL_MACHINE\SOFTWARE\Usoris]
> 
> [HKEY_LOCAL_MACHINE\SOFTWARE\Usoris\Remote Utilities Host]
> 
> [HKEY_LOCAL_MACHINE\SOFTWARE\Usoris\Remote Utilities Host\Host]
> ```
> 
> So the beginning of the file must look like:
> 
> ```
> Windows Registry Editor Version 5.00
> 
> [HKEY_LOCAL_MACHINE\SOFTWARE\Usoris\Remote Utilities Host\Host\Parameters]
> ```

ok, I'll do it in a couple of hours

yes1n committed:
Update rms-7.0.0.1-Parameters.reg

#499 https://github.com/glpi-project/glpi-agent/pull/499#pullrequestreview-1648927438

yes1n committed:
Merge branch 'RMS_MGMT' of https://github.com/yes1n/glpi-agent into RMS_MGMT

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm:
```diff
@@ -0,0 +1,74 @@
+    my $inventory = $params{inventory};
+    my $logger    = $params{logger};
+
+    my $InternetID = _getID(logger  => $logger);
+    
+    if (defined($InternetID)) {
```
There's still a issue I missed here, empty lines should be empty:
```suggestion
    my $InternetID = _getID(logger  => $logger);

    if (defined($InternetID)) {
```

g-bougard reviewed (changes requested):
Hi @yes1n 
sorry, the AUTHOR tests failed again. This time on remaining spaces on a line that should remain empty.
After that, your PR should be fine for merging.

yes1n committed:
Update lib/GLPI/Agent/Task/Inventory/Generic/Remote_Mgmt/RMS.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

yes1n commented:
@g-bougard Thank you for approving the idea and helping me write food.  I would even say that you did almost everything.  I even thought about learning Perl.
